### PR TITLE
Integrate Time Series Classification Pipelines into AutoMLSearch

### LINKED ISSUE #1580: Integrate time series classification pipelines into AutoMLSearch

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1580

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/tests/automl_tests/test_automl.py [hunk 1]
```diff
@@ -1976,32 +1972,6 @@ def test_automl_validates_problem_configuration(X_y_binary):
     assert problem_config == {"max_delay": 2, "gap": 3}
 
 
-@patch('evalml.pipelines.TimeSeriesRegressionPipeline.score', return_value={"R2": 0.3})
```
Comment (freddyaboulton) [inline evalml/tests/automl_tests/test_automl.py:1979]: Moving this test to `test_automl_search_regression.py`

File: evalml/pipelines/utils.py [hunk 2]
```diff
@@ -67,7 +71,7 @@ def _get_preprocessing_components(X, y, problem_type, text_columns, estimator_cl
     if add_datetime_featurizer:
         pp_components.append(DateTimeFeaturizer)
 
-    if problem_type in [ProblemTypes.TIME_SERIES_REGRESSION]:
+    if is_time_series(problem_type):
```
Comment (freddyaboulton) [inline evalml/pipelines/utils.py:74]: So that AutoML can create pipelines for ts classification when `allowed_pipelines=None`

File: evalml/pipelines/time_series_baselines.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from evalml.pipelines import (
+    TimeSeriesBinaryClassificationPipeline,
+    TimeSeriesMulticlassClassificationPipeline,
+    TimeSeriesRegressionPipeline
+)
+
+
+class TimeSeriesBaselineRegressionPipeline(TimeSeriesRegressionPipeline):
```
Comment (freddyaboulton) [inline evalml/pipelines/time_series_baselines.py:8]: Consolidating all of the baseline pipelines for ts into the same file

File: evalml/tests/pipeline_tests/test_time_series_pipeline.py [hunk 4]
```diff
@@ -322,36 +296,6 @@ class Pipeline(pipeline_class):
     pl.score(X, y, objectives)
 
 
-@pytest.mark.parametrize("pipeline_class", [TimeSeriesBinaryClassificationPipeline,
-                                            TimeSeriesMulticlassClassificationPipeline])
-@pytest.mark.parametrize("use_none_X", [True, False])
-def test_score_works_with_estimator_uses_y(use_none_X, pipeline_class, X_y_binary, X_y_multi):
```
Comment (freddyaboulton) [inline evalml/tests/pipeline_tests/test_time_series_pipeline.py:328]: We have coverage for this case in `test_time_series_baseline_pipeline.py`

File: evalml/automl/automl_search.py [hunk 5]
```diff
@@ -634,10 +636,14 @@ def _add_baseline_pipelines(self):
         elif self.problem_type == ProblemTypes.REGRESSION:
             baseline = MeanBaselineRegressionPipeline(parameters={})
         else:
+            pipeline_class = {ProblemTypes.TIME_SERIES_REGRESSION: TimeSeriesBaselineRegressionPipeline,
```
Comment (chukarsten) [inline evalml/automl/automl_search.py:639]: I like this.  It makes me feel good in a python kind of way.

File: evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py [hunk 6]
```diff
@@ -6,19 +6,21 @@
...
+    pad_with_nans
 )
 
 
-class TimeSeriesBaselineRegressor(Estimator):
+class TimeSeriesBaselineEstimator(Estimator):
     """Time series regressor that predicts using the naive forecasting approach.
 
     This is useful as a simple baseline regressor for time series problems
     """
```
Comment (chukarsten) [inline evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py:18]: Does this need to change too?

File: evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py [hunk 7]
```diff
@@ -63,6 +65,16 @@ def predict(self, X, y=None):
 
         return y
 
+    def predict_proba(self, X, y=None):
+        if y is None:
+            raise ValueError("Cannot predict Time Series Baseline Regressor if y is None")
```
Comment (chukarsten) [inline evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py:70]: Regressor -> Estimator?

File: evalml/tests/pipeline_tests/test_pipelines.py [hunk 8]
```diff
@@ -51,8 +51,7 @@
     make_pipeline,
     make_pipeline_from_components
 )
-from evalml.preprocessing.utils import is_time_series
-from evalml.problem_types import ProblemTypes
+from evalml.problem_types import ProblemTypes, is_time_series
```
Comment (chukarsten) [inline evalml/tests/pipeline_tests/test_pipelines.py:54]: That's probably a smart move.

**Review by chukarsten — CHANGES_REQUESTED** [review]
Just some small, comment related stuff.  Looks good though.  Hefty, hefty.

File: evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py [hunk 9]
```diff
@@ -63,6 +65,16 @@ def predict(self, X, y=None):
...
+        y = _convert_to_woodwork_structure(y)
+        y = _convert_woodwork_types_wrapper(y.to_series())
+        preds = self.predict(X, y).dropna(axis=0, how='any').astype('int')
+        proba_arr = np.zeros((len(preds), y.max() + 1))
+        proba_arr[np.arange(len(preds)), preds] = 1
+        return pad_with_nans(pd.DataFrame(proba_arr), len(y) - len(preds))
+
     @property
     def feature_importance(self):
         """Returns importance associated with each feature. Since baseline regressors do not use input features to calculate predictions, returns an array of zeroes.
```
Comment (ParthivNaresh) [inline evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py:80]: Might need to change regressors to estimators here as well

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 10]
```diff
@@ -111,7 +111,11 @@ def predict(self, X, y=None, objective=None):
         y = _convert_woodwork_types_wrapper(y.to_series())
         n_features = max(len(y), X.shape[0])
         predictions = self._predict(X, y, objective=objective, pad=False)
-        predictions = pd.Series(self._decode_targets(predictions), name=self.input_target_name)
+
+        # In case gap is 0 and this is a baseline pipeline, we drop the nans in the
+        # predictions before decoding them
+        predictions = pd.Series(self._decode_targets(predictions.dropna()), name=self.input_target_name)
```
Comment (ParthivNaresh) [inline evalml/pipelines/time_series_classification_pipelines.py:117]: Good catch!

**Review by ParthivNaresh — APPROVED** [review]
Looks good to me! Just a few docstring updates

File: evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py (hunk 6 above)
Comment (freddyaboulton) [reply evalml/pipelines/components/estimators/regressors/time_series_baseline_estimator.py:18]: Definitely!

File: evalml/tests/pipeline_tests/test_time_series_baseline_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+@pytest.mark.parametrize('pipeline_class', [TimeSeriesBaselineRegressionPipeline,
+                                            TimeSeriesBaselineBinaryPipeline, TimeSeriesBaselineMulticlassPipeline])
+def test_time_series_baseline(pipeline_class, gap, X_none, ts_data):
+    X, y = ts_data
+
+    clf = pipeline_class(parameters={"pipeline": {"gap": gap, "max_delay": 1},
+                                     "Time Series Baseline Estimator": {'gap': gap, 'max_delay': 1}})
+    expected_y = y
+    if gap == 0:
+        expected_y = y.shift(1)
```
Comment (bchen1116) [inline evalml/tests/pipeline_tests/test_time_series_baseline_pipeline.py:25]: could shorten this to 
```python
expected_y = y.shift(1) if gap == 0 else y
```

**Review by bchen1116 — COMMENTED** [review]
Not sure if this is supposed to work currently, but 
image

Additionally, something that we didn't catch previously, but we should change the datachecks we use when we are doing time series problems, specifically the `class_imbalance_data_check` since `TimeSeriesSplit` doesn't necessarily shuffle/stratify the data that the other data splitters do: 
image
This can definitely be filed as a separate issue.

The rest looks good to me!

**Comment by freddyaboulton** [comment]
@bchen1116 I filed #1681 to track updates to the class imbalance datacheck. The issue with the fraud dataset is closely related to #1507 because the problem is that the pipelines are delaying categorical features (which introduces NaNs) and then trying to OHE the delayed categorical columns (which can't handle NaNs). I think the problem is that we need a way of identifying which features depend on time and which don't: For fraud we may want to have a record of the currencies used in the last k transactions, but we shouldn't delay features which don't depend on time (like the customer_id if present). I'll add some notes to #1507, we may have to file a separate issue.

**Review by chukarsten — APPROVED** [review]
